### ISSUE TITLE
Valid refresh token gives error

### ISSUE THREAD

**Issue by rajeevBlr** [keycloak/keycloak#14040]
### Describe the bug

Context:
We are using onTokenExpired event of Keycloak from 'keycloak-js'  to refresh the access token upon expiry. 
We call updateToken method when onTokenExpired is fired.

Issue:
This works mostly without any issues. But randomly we get 400 error when we try to refresh the token. 
In our realm, access_token expires in 5 minutes and refresh_token expires in 30 minutes.
The updateToken calls https://abc.xyz.com/realms/my_realm/protocol/openid-connect/token with grant_type=refresh_token 
It gives a new access_token and a NEW refresh_token each time it's called.

Subsequent updateToken call uses the refresh_token from previous request.
So in our setup access_token expires every 5 minutes -> updateToken called every 5 minutes -> We can new refresh_token every 5 minutes.

This works for hours and is expected to work till "SSO Session Max" which is set to 10 hours.

But randomly (can be after 2 hours or after 3-4 hours), https://abc.xyz.com/realms/my_realm/protocol/openid-connect/token gives 400 error. Message logged by KeyCloak around same time is:

WARN  [org.keycloak.events] (executor-thread-246) type=REFRESH_TOKEN_ERROR, realmId=my_realm, clientId=client-web, userId=xxx, ipAddress=xyz error=invalid_token, grant_type=refresh_token, refresh_token_type=Refresh, refresh_token_id=8649a5b1-ca82-4908-a649-90e517050af6, client_auth_method=client-secret

The puzzle is:

1. We used a refresh_token  which was created just 5 minutes back with validity of 30 minutes.
2. So refresh_token has not expired, then why do we get 400 error with no clear reason?

### Version

17.0.1

### Expected behavior

updateToken should be able to refresh access_token till "SSO Session Max" period which is set to 10 hours for our case.

### Actual behavior

A nonexpired refresh_token is treated as invalid_token

### How to Reproduce?

Maybe use the same setting as mine:

SSO Session Idle: 30 minutes
SSO Session Max: 10 hours
Access Token Lifespan: 5 minutes

**Comment by lexcao** [comment]
> But randomly (can be after 2 hours or after 3-4 hours), `../token` gives 400 error.

> So in our setup access_token expires every 5 minutes -> updateToken called every 5 minutes -> We can new refresh_token every 5 minutes.

Yes, as you said, it works for hours and is expected to work till SSO session max.

There may be some client behaviors could prevent the `keycloak.js` regular refreshing token.
* laptop closing the lid
* computer sleeping 
* ...other may prevent js run in background

For example,
```
12:00 call `/token` to get the access token and refresh token
12:05 access token should be expired and refresh token is called

...
13:00 ~ 14:00 may be there is no activity in the application page 
                        or the computer is sleeping, no refresh token is called
...

14:40 after back to active, the keycloak.js notices the access token is expred so to refresh token but it is expired as well
```

I think it is an expected error and no need worry about that.

**Comment by rajeevBlr** [comment]
Hi @lexcao thanks for responding.

In my case, the client computer was active through out the period.
Every 5 minutes we were able to get new access token and new refresh token using a refresh token. 

Since refresh token is valid for 30 minutes and we get new refresh token every 5 minutes, I don't understand why refresh token will be considered expired in random instances.

**Comment by lexcao** [comment]
Hi, another guess,
how many tabs have you opened?

* BTW, what is your `Revoke Refresh Token` setting?

some reference https://github.com/keycloak/keycloak/discussions/10937

**Comment by rajeevBlr** [comment]
Thanks again!

`Revoke Refresh Token` is turned off in my realm.
We could replicate this 400 error in single tab too.

**Comment by gabrielsch** [comment]
We are also facing this issue in our apps -- @rajeevBlr were you able to figure this out?

Also, what is the proper handling once the SSO Session Max expires? Do a full page refresh?

**Comment by mposolda** [comment]
This looks like a valid issue, but not a blocker for Keycloak 22. Adding to backlog for now.

It will be good to have more details how it can happen as currently it seems the issue is non-deterministic, which makes it hard to nail down and fix...

**Comment by daviddelannoy** [comment]
Hi,

@mposolda : please find below some context data.
Same issue observed here with this configuration : 

- KC **20.0.5**
- `Revoke refresh token` : off
- `SSO Session Idle` : 30 days 
- `SSO Session Max` : several weeks
- `Access token lifespan` : 10 min
- Access token lifespan for implict flow : 15 min 

Tests were done on **Android** (version does not matter, we can reproduce with Android 8 to 13) with **Google Chrome as default browser**. 
`web-app-1` and `web-app-2` are two websites visited with Google Chrome and `mobile-app` is a **native Android App**, using [AppAuth-Android](https://github.com/openid/AppAuth-Android) last version and Chrome custom tabs.

Web Apps are using Keycloak JS adapter up-to-date with : 

- `onLoad:"check-sso"`
- `flow:"implicit flow"` (sorry ...)
- `silentCheckSsoFallback: false,`
- a valid `silentCheckSsoRedirectUri` 

SSO works fine in any browser between these apps and other ones, but not with the Android App. 
Regularly, we are disconnected from `mobile-app` because of the same `REFRESH_TOKEN_ERROR` : 

```
"type=CODE_TO_TOKEN, realmId=test, clientId=mobile-app, userId=0575c5b3-0ead-4dba-b1e9-59a667a7ddd1, ipAddress=34.164.209.117, token_id=06783d52-dbe4-49f2-8b30-a2e783e76ebc, grant_type=authorization_code, refresh_token_type=Refresh, scope='openid email profile', refresh_token_id=a0494468-f6dc-4d0b-b137-f1182a574e01, code_id=d97d6fd1-f797-4050-97b7-a3537522bfe4, client_auth_method=client-secret"
"type=REFRESH_TOKEN_ERROR, realmId=test, clientId=mobile-app, userId=0575c5b3-0ead-4dba-b1e9-59a667a7ddd1, ipAddress=34.164.209.117, error=invalid_token, grant_type=refresh_token, refresh_token_type=Refresh, refresh_token_id=a0494468-f6dc-4d0b-b137-f1182a574e01, client_auth_method=client-secret"
```

This `REFRESH_TOKEN_ERROR` will always follow either a successful REFRESH_TOKEN in a Google Chrome tab for `web-app-1`  :  

`"type=REFRESH_TOKEN, realmId=test, clientId=web-app-1, userId=0575c5b3-0ead-4dba-b1e9-59a667a7ddd1, ipAddress=47.249.172.57, token_id=1f5daeca-fee8-412c-a2d4-2e89324fb2ac, grant_type=refresh_token, refresh_token_type=Refresh, updated_refresh_token_id=4bc42692-e1fb-4fbc-a754-61e88af9bc26, scope='openid', refresh_token_id=fa8ab4d5-227c-4d8d-b834-abb545a3b962, client_auth_method=client-secret"`

or a login with SSO on `web-app-2`  in another tab on for example : 

`"type=LOGIN, realmId=test, clientId=web-app-2, userId=0575c5b3-0ead-4dba-b1e9-59a667a7ddd1, ipAddress=34.164.209.117, auth_method=openid-connect, auth_type=code, response_type=code, redirect_uri=https://another-redirect-uri, authSessionTabId=Rc6eatWHV8E"`

Regarding code of [TokenManager, session is not valid](https://github.com/keycloak/keycloak/blob/20.0.5/services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java#L152) because `userSession` is null.

Which can be noticed easily in Keycloak admin console as the previous existing user session on mobile-app just disapear and is replaced by a totally new one on other client ids ... 

Just after login in the app for a new user Session  : 
![image](https://github.com/keycloak/keycloak/assets/16318239/1e65cecc-6084-46a1-bc0c-2b318b36205e)

Then after navigating in SSO to other apps in Google Chrome tabs :  
![image](https://github.com/keycloak/keycloak/assets/16318239/7730c87c-4419-4028-afa4-4e6fd1d01692)

For sure on the next refresh_token (10 min max later) the `invalid_grant` with `Session not active` is triggered.
Just like the user navigated to `web-app-1` and `web-app-2` with `check-sso`, he just has to click on "Log in" in the `mobile-app` and is logged in again ... (thanks to an identity provider on which he still has an active session & cookie and used in a mandatory identity provider redirector).

But it is quite annoying as he will always get a popup telling him that he was disconnected which is technically true on keycloak User & Client session's side. 

I think that we started to observe a increasing volume of "Session not active" after upgrading from Keycloak 17 to 19.0.3, and still present in 20.0.5

We were not able to identify what can be specific to the Android App that clear the client an user session and does not occur between two web apps accessed by a browser. This scenario leads easily to the problem mentioned before, but we may also have the random `REFRESH_TOKEN_ERROR` too on a single app.

**Comment by Tristan971** [comment]
This has been a recurring issue on our end too, so we figured we might share some extra data.

Like others, the observed issue on our end is that refresh tokens get randomly expired, much quicker than should be the case given our session/token expiry configuration. For example, we upgrade to Keycloak 21.1.1 a few days ago, and have 28 days expiry across the board, yet we still have 3-5 refreshes per second failing with those:

```
2023-07-11 10:14:04,584 WARN  [org.keycloak.events] (executor-thread-62335) type=REFRESH_TOKEN_ERROR, realmId=mangadex, clientId=mangadex-frontend-stable, userId=c1e08da4-1191-4643-adf0-bd87fc56fbf6, ipAddress=redacted, error=invalid_token, grant_type=refresh_token, refresh_token_type=Refresh, refresh_token_id=c1b70219-868a-47aa-87e6-6df0738ae1c5, client_auth_method=client-secret
2023-07-11 10:14:05,254 WARN  [org.keycloak.events] (executor-thread-14015) type=REFRESH_TOKEN_ERROR, realmId=mangadex, clientId=mangadex-frontend-stable, userId=26d855a9-4450-4fb5-896e-f5760ff7528c, ipAddress=redacted, error=invalid_token, grant_type=refresh_token, refresh_token_type=Refresh, refresh_token_id=e4e04e96-0086-46c6-9b1c-d0938eeb5838, client_auth_method=client-secret
2023-07-11 10:14:05,258 WARN  [org.keycloak.events] (executor-thread-44708) type=REFRESH_TOKEN_ERROR, realmId=mangadex, clientId=mangadex-frontend-stable, userId=null, ipAddress=98.128.213.15, error=invalid_token, grant_type=refresh_token, client_auth_method=client-secret
2023-07-11 10:14:05,375 WARN  [org.keycloak.events] (executor-thread-62331) type=REFRESH_TOKEN_ERROR, realmId=mangadex, clientId=mangadex-frontend-stable, userId=6ac04392-4d5a-469e-b4a2-90ead48314a0, ipAddress=redacted, error=invalid_token, grant_type=refresh_token, refresh_token_type=Refresh, refresh_token_id=43a65ba3-15cf-472c-908a-1e477843593d, client_auth_method=client-secret
2023-07-11 10:14:05,604 WARN  [org.keycloak.events] (executor-thread-62331) type=REFRESH_TOKEN_ERROR, realmId=mangadex, clientId=mangadex-frontend-stable, userId=null, ipAddress=redacted, error=invalid_token, grant_type=refresh_token, client_auth_method=client-secret
2023-07-11 10:14:07,034 WARN  [org.keycloak.events] (executor-thread-14008) type=REFRESH_TOKEN_ERROR, realmId=mangadex, clientId=mangadex-frontend-stable, userId=440cb483-451c-410e-ad5c-06d6b8296353, ipAddress=redacted, error=invalid_token, grant_type=refresh_token, refresh_token_type=Refresh, refresh_token_id=dc73f65b-4c5f-4ef0-b8af-e3c527552048, client_auth_method=client-secret
2023-07-11 10:14:07,732 WARN  [org.keycloak.events] (executor-thread-62326) type=REFRESH_TOKEN_ERROR, realmId=mangadex, clientId=mangadex-frontend-stable, userId=null, ipAddress=redacted, error=invalid_token, grant_type=refresh_token, client_auth_method=client-secret
2023-07-11 10:14:07,801 WARN  [org.keycloak.events] (executor-thread-14008) type=REFRESH_TOKEN_ERROR, realmId=mangadex, clientId=mangadex-frontend-stable, userId=null, ipAddress=redacted, error=invalid_token, grant_type=refresh_token, client_auth_method=client-secret
2023-07-11 10:14:07,962 WARN  [org.keycloak.events] (executor-thread-14008) type=REFRESH_TOKEN_ERROR, realmId=mangadex, clientId=mangadex-frontend-stable, userId=null, ipAddress=redacted, error=invalid_token, grant_type=refresh_token, client_auth_method=client-secret
2023-07-11 10:14:08,878 WARN  [org.keycloak.events] (executor-thread-14008) type=REFRESH_TOKEN_ERROR, realmId=mangadex, clientId=mangadex-frontend-stable, userId=null, ipAddress=redacted, error=invalid_token, grant_type=refresh_token, client_auth_method=client-secret
```

The first interesting piece is that these errors follow 2 categories:
1. Sometimes userId = null
2. Sometimes userId != null

And it's about evenly shared between the two cases; so it doesn't look like that's important here.

Looking further into our logs (we use PostgreSQL and embedded Infinispan+k8s discovery), we do see quite a few infinispan-related warnings every couple of minutes, which seem suspiciously related to the issue of sessions getting lost:
```
2023-07-11 10:02:08,005 WARN  [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (executor-thread-44655) Failed to replace entity '67f61c04-9a4f-4b09-81c2-05f7f9f990ad' in cache 'sessions'
2023-07-11 10:02:08,029 WARN  [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (executor-thread-11746) Failed to replace entity '67f61c04-9a4f-4b09-81c2-05f7f9f990ad' in cache 'sessions'
2023-07-11 10:02:08,032 WARN  [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (executor-thread-11757) Failed to replace entity '67f61c04-9a4f-4b09-81c2-05f7f9f990ad' in cache 'sessions'
2023-07-11 10:02:08,039 WARN  [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (executor-thread-62257) Failed to replace entity '67f61c04-9a4f-4b09-81c2-05f7f9f990ad' in cache 'sessions'
2023-07-11 10:02:08,048 WARN  [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (executor-thread-13949) Failed to replace entity '67f61c04-9a4f-4b09-81c2-05f7f9f990ad' in cache 'sessions'
2023-07-11 10:02:08,050 WARN  [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (executor-thread-11758) Failed to replace entity '67f61c04-9a4f-4b09-81c2-05f7f9f990ad' in cache 'sessions'
2023-07-11 10:02:25,598 WARN  [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (executor-thread-11746) Failed to replace entity '51363e9d-ff84-47ca-9f0b-ad205334f456' in cache 'sessions'
2023-07-11 10:02:25,622 WARN  [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (executor-thread-11759) Failed to replace entity '51363e9d-ff84-47ca-9f0b-ad205334f456' in cache 'sessions'
2023-07-11 10:02:25,660 WARN  [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (executor-thread-44656) Failed to replace entity '51363e9d-ff84-47ca-9f0b-ad205334f456' in cache 'sessions'
2023-07-11 10:03:54,296 WARN  [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (executor-thread-13949) Failed to replace entity '1328551e-f813-4d60-9079-bade450008cf' in cache 'sessions'
2023-07-11 10:03:54,338 WARN  [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (executor-thread-44668) Failed to replace entity '1328551e-f813-4d60-9079-bade450008cf' in cache 'sessions'
2023-07-11 10:03:54,344 WARN  [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (executor-thread-44673) Failed to replace entity '1328551e-f813-4d60-9079-bade450008cf' in cache 'sessions'
2023-07-11 10:03:54,357 WARN  [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (executor-thread-11756) Failed to replace entity '1328551e-f813-4d60-9079-bade450008cf' in cache 'sessions'
2023-07-11 10:03:54,353 WARN  [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (executor-thread-62292) Failed to replace entity '1328551e-f813-4d60-9079-bade450008cf' in cache 'sessions'
2023-07-11 10:03:54,356 WARN  [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (executor-thread-62287) Failed to replace entity '1328551e-f813-4d60-9079-bade450008cf' in cache 'sessions'
2023-07-11 10:03:54,394 WARN  [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (executor-thread-62286) Failed to replace entity '1328551e-f813-4d60-9079-bade450008cf' in cache 'sessions'
2023-07-11 10:03:54,401 WARN  [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (executor-thread-62296) Failed to replace entity '1328551e-f813-4d60-9079-bade450008cf' in cache 'sessions'
2023-07-11 10:03:54,405 WARN  [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (executor-thread-62297) Failed to replace entity '1328551e-f813-4d60-9079-bade450008cf' in cache 'sessions'
2023-07-11 10:03:54,408 WARN  [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (executor-thread-13951) Failed to replace entity '1328551e-f813-4d60-9079-bade450008cf' in cache 'sessions'
2023-07-11 10:03:54,408 WARN  [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (executor-thread-62299) Failed to replace entity '1328551e-f813-4d60-9079-bade450008cf' in cache 'sessions'
2023-07-11 10:03:54,409 WARN  [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (executor-thread-13955) Failed to replace entity '1328551e-f813-4d60-9079-bade450008cf' in cache 'sessions'
2023-07-11 10:03:54,411 WARN  [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (executor-thread-13957) Failed to replace entity '1328551e-f813-4d60-9079-bade450008cf' in cache 'sessions'
```

Now we first thought the problem was a lack of memory for Keycloak's embedded infinispan to hold all of our sessions (going off client-session-stats, we have around 470k active sessions right now), so we raised the memory per KC instance to 6GB at the container level and 3GB of JVM heap. With this, we get the following (average) memory reports:
![image](https://github.com/keycloak/keycloak/assets/875533/88f6c66f-e57b-4e40-98d4-623ff7835976)

So things are either just fine (we observe approx 4/6GB of actual memory usage at the container level) or heap usage is misrepresented (maybe it doesn't account for the embedded infinispan at all? I don't know).

Another thing we noticed is that occasionally we'd observe Keycloak suddenly rejecting 100% of refresh tokens (basically, the session cache got completely evicted). This never (or almost never at least) happens on its own, however whenever we had some infrastructure strain (CPU strain, FS stalls, network hiccups, ... you name it), that is quite likely to happen. Essentially logging out all of our users at once.

Finally, Keycloak just seems to have difficulties overall with handling scale-ups and scale-downs (we run 5-8 instances on average, oscillating over time), unless they are extremely carefully ordered (1-by-1 with long waiting times between them). My guess is that this is more related to Infinispan itself, and we're looking at options for deploying our own external one that we could observe much better than the embedded one... Unfortunately the documentation on the matter is sparse, to say the least.

Hopefully this gives a few hints chasing this issue.

**Comment by dglozano** [comment]
We have experienced the same as @Tristan971 with an identical setup. 4 nodes with 8 gb each and 6gb of heap size. They are utilizing ~3GB each. Cache owners is set to 4. 

The only thing I could add is that it seems to happen whenever we get a sudden burst of calls to the refresh token endpoint.
Also, is the `Failed to replace entity` errors happen both for `sessions` cache and `clientSessions` cache.

**Comment by todorinskiz** [comment]
While the user token is valid for 2 minutes, our client settings had Client Session Idle set to 1 minute:
![image](https://github.com/keycloak/keycloak/assets/6491638/8dc38e03-aec4-4d29-b874-2e5ffd8a8d2b)

This means that the session of our Spring Cloud Gateway (as authentication gateway, aka the client) was idle for 2 minutes, and presumably Keycloak would invalidate the session, rendering the refresh tokens from that session invalid.

After clearing the value of the Client Session Idle it falls back to SSO Session Idle value.
This one you can find in the realm token settings have SSO Session idle, and ours is set to 2 hours.
![image](https://github.com/keycloak/keycloak/assets/6491638/d603fa68-9f18-400e-a104-936541a31465)

**Comment by Tristan971** [comment]
While this method works around… other questionable decisions in the way tokens unexpectedly expire, it is a different thing and not a solution to this issue.

We are aware that a token will never outlive its associated session, and we do (like you suggest) make sure that the underlying session does NOT expire faster.

Yet the problem reported remains.
Frankly, from the logs I shared and the times when it suddenly manifests much more frequently (during compute strain episodes), it smells like a long standing concurrency bug somewhere, whether in the glue code between KC and Infinispan, or within Infinispan itself.

**Comment by daviddelannoy** [comment]
Like you said it @Tristan971 , we also noticed that ISPN cache errors (essentially ISPN000936) might occur more frequently or easily when adding/removing some new nodes to the cluster.
When doing rolling updates, it is much more stable to increase and decrease one by one and waiting a bit (one minute) than cache replication occured and finished correctly, even with "owners=2" and distributed-cache.
We also noticed an improvement since Quarkus (almost divided by 2 the duration of a full rolling on 3 instances), essentially thanks to the quicker startup time.

I also think that these unexpected ISPN errors on user sessions updates/replacement may be a root cause for too soon refresh_token_errors, because it prevents from finding the valid attached user session. 

It would be interesting to have the same setup with an ISPN external cluster to see if it is reproducible.

**Comment by Tristan971** [comment]
> It would be interesting to have the same setup with an ISPN external cluster to see if it is reproducible.

Funny you mention that, I very strongly considered that approach too, and I suspect it would greatly help so long as the external infinispan isn't subjected to the same level of autoscaling that KC itself might.

I haven't gone through however, partly due to the lack of documentation on the matter (or my inability to find it, who knows), and partly because this issue doesn't seem to be getting the attention I feel it should warrant, making me wonder if that is a good investment of my time to research...

**Comment by daviddelannoy** [comment]
I might have some configuration inputs for you as we already tried with ispn 13.x, succeeded but then faced expiration issue [10755](https://github.com/keycloak/keycloak/issues/10755) that has been fixed in the meantime. 
It was with the legacy wildfly version so some adjustments may be required for KC 22.x
Never had the chance to set up the full external cluster since but it should work. 

 Will try to look into archive next week 😉

**Comment by torsorensen** [comment]
Hey, 

I want to mention we are facing the same random REFRESH_TOKEN_ERROR behaviour as explained in this thread. We've never managed to figure out why it happens, so I'd like to contribute as good as possible to this investigation.

What is particularly interesting to add from our setup is that we were on KC 13.0.0 (WildFly) for quite a long time, and we were ALWAYS having 1 pod running in Kubernetes only, and thus ISPN was never configured. Never. This is quite the opposite of everyone in this thread so far.

Now since recently, we've upgraded to KC 22.0.5 (Quarkus) and we're running in 2-3 pod setup with ISPN and Kubernetes cache stack on.

And in both setups (KC13 and KC22) we have been consistently seeing the warning daily in our logs.

We haven't paid attention to increased warnings during stress, so can't comment on that, but we do see the mix of userId = null and userId != null as already mentioned earlier in this thread.

The only minor difference we've found so far between KC 13 and KC 22 is that on KC 13 the logs were coloured red in our logging tool, and now they're coloured white. But I don't want to attribute this any significance.

Token-wise, we have some different configurations across different applications
- One application uses the Code flow and has access token 15 mins, refresh token 30 mins, with a max of 6 hours. 
- Another application uses Direct flow (simply password/username) with access token 15 mins and offline token (i.e. refresh token without expiry)

I hope this can provide additional value in the pursuit of this tricky bug.

**Comment by bekwam** [comment]
I replicated this on 21.1.2.  I set SSO Session Idle to 5m.  I make a successful refresh_token call at 4m and one at 8m.  The call at 8m fails with 400 error=invalid_grant, error_description=Token is not active.

The refresh call is POSTed to the /token endpoint with a URL-encoded payload that follows

```
"grant_type=refresh_token&client_id="  + client + "&refresh_token=" + refreshToken
```

It seems like the counter tracking session refresh expiration is not reset.

This was with a user created in KC (no user federation) and with idle settings made at both the realm and the client level.  My workaround is to increase the idle time to that of a workday (10 hrs).  However, if someone is working a long day with continuous usage, they'll get a timeout.

**Comment by Tristan971** [comment]
> I set SSO Session Idle to 5m.
> I make a successful refresh_token call at 4m and one at 8m.
> The call at 8m fails with 400 error=invalid_grant, error_description=Token is not active.

That one is a slightly different issue. Namely that token refreshes do not count towards keeping the session active... Or at least not consistently.

I also found that extremely strange at first but maybe it's by design. Either way, we raised our idle timeout to be equal to our active one and the issue remained.

The root cause of this issue can be summed up as:
- sessions are only stored in memory in infinispan
- infinispan loses that data at times (it *seems* to be when handling joining/leaving peers, but we don't have proof of that)

**Comment by ferdoran** [comment]
We are also observing this behaviour regularly, especially on Android platforms. Unfortunately this forces customers to reauthenticate within active sessions. We're on Keycloak 21.1.1 with an external Infinispan.

Here's my 2 cents on the session data loss:

Sessions are not just stored in Infinispan, but also in the database configured with Infinispan. If in-memory cache data is lost for whatever reason, it should be loaded from the database instead. Losing the session data completely (also in the database) is something that should only happen when the session expires or is closed (e.g. by logout) and not when in-memory data is lost.

**Comment by PhungVanHoa** [comment]
This problem also occurs in version 23.0.1.

**Comment by Tristan971** [comment]
For what it's worth, about 2 months ago we changed our settings to never autoscale Keycloak, and the issue has almost entirely disappeared.

So the root cause is definitely poor handling of joining/leaving cluster members (whether in KC or in Infinispan).

**Comment by saurabhnemade** [comment]
We are observing this issue. I have written middleware in nextjs to refresh token when accesstoken is invalid or expired and I am logging all values of refresh token in datadog. 

I can clearly see that refresh token's `exp` is 3 months in future but token fails to do refresh call. On datadog for keycloak instance I see

 `type=REFRESH_TOKEN_ERROR, realmId=<REDACTED>, clientId=web, userId=f:<REDACTED>:<REDACTED>, ipAddress=<REDACTED>, error=invalid_token, grant_type=refresh_token, refresh_token_type=Refresh, refresh_token_id=3a998a97-5a9e-4a14-96c6-c0a4d3d38b3b, client_auth_method=client-secret`

**Comment by saurabhnemade** [comment]
@stianst Any ETA on when it will be picked up next for fixing? 
It's very common to use OAuth solutions in kubernetes env where multiple nodes are created and if cache can't be fixed immediately, can you suggest any other way of handling this? Perhaps by showing us a way of using Redis as cache with keycloak?

**Comment by Tristan971** [comment]
> can you suggest any other way of handling this?

Well you have three options now:
1. Not autoscaling keycloak at all anymore (like I did)
2. Deploying your own Infinispan cluster, externally to keycloak, and have KC use that (but how to deploy/configure Infinispan for KC as client is not documented I think)
3. Making sure your autoscaling in k8s is very *slow* (like @daviddelannoy [noted in this comment](https://github.com/keycloak/keycloak/issues/14040#issuecomment-1694232849))

> When doing rolling updates, it is much more stable to increase and decrease one by one and waiting a bit (one minute) [...]

The funny thing is that it's actually not easy to achieve consistently in k8s for all situations (manual or HPA-driven, scale up or down, spec changes rollout, deployment vs statefulset). But this is what I would do if I reaaaaallly needed autoscaling.

First, you want to add policies to HPA(s) looking something like this:
```yaml
apiVersion: autoscaling/v2
kind: HorizontalPodAutoscaler
metadata:
  name: keycloak
spec:
  ...
  behavior:
    scaleUp:
      ...
      policies:
        - type: Pods
          value: 1
          periodSeconds: 60
    scaleDown:
      ...
      policies:
        - type: Pods
          value: 1
          periodSeconds: 60
```

This should ensure that HPAs will refuse to increase/decrease replica count by more than one pod per minute. (if you don't use native HPAs but instead use something like Argo Rollouts... good luck!)

But this is not enough; because this applies when both of these conditions are met:
- it is scale up/down event (rather than a new deployment/statefulset spec being pushed)
- it is triggered by this HPA (and not by a human editing the spec replica count)

For these, if you use a Deployment:
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: "keycloak"
spec:
  strategy:
    type: RollingUpdate
    rollingUpdate:
      maxUnavailable: 1
      maxSurge: 1
  minReadySeconds: 60
```

And if a StatefulSet, like this:
```
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: keycloak
spec:
  updateStrategy:
    type: RollingUpdate
    rollingUpdate:
      maxUnavailable: 1
  minReadySeconds: 60
  podManagementPolicy: OrderedReady
```

But even these don't actually ensure that scale-down is at most 1 per minute (without heavy trickery like lifecycle preStop scripts maybe).

Also I could swear I saw k8s ignore `maxSurge` at times so YMMV (I'm sure there was a reason, but it surprised me in the past).

So yeah, a bit shit and why I personally just stuck the max amount I need on a daily basis and disabled scale-up/down. But my use-case sees at most 30% traffic difference between daily minimum and maximum, so it's not as bad as for others...

**Comment by helielson** [comment]
@stianst any updates here? I noticed you removed this issue from the backlog.

**Comment by CaptainDonut** [comment]
@stianst We're still seeing this issue as well, any updates?

**Comment by saurabhnemade** [comment]
@stianst I have found one more case when session loss occurs.

When same user is logged in from multiple browsers and one clicks on logout it clears all sessions from keycloak instance!!

**Comment by infl00p** [comment]
We are experience this problem even with an external Infinispan cluster. Could it be a variant of the https://github.com/keycloak/keycloak/issues/27117 that seems to be resolved in KC 24.0.2? This one was about revoking refresh tokens, but actually about an infinispan issue with the caches.
We are using KC 23.0.7

**Comment by ahus1** [comment]
> I am redirecting user to logout url but before that I am calling token endpoint to revoke both access_token and refresh_token.

@saurabhnemade - revoking the refresh token might cause the the effect you see. While this was unexpected to me as well at first, it is in the [OAuth 2.0 specs](https://datatracker.ietf.org/doc/html/rfc7009): 

> If the particular
   token is a refresh token and the authorization server supports the
   revocation of access tokens, then the authorization server SHOULD
   also invalidate all access tokens based on the same authorization
   grant

When then in one browser the refresh token is revoked for that user for that client, it will remove the client session from all other user sessions of that user. If then no other client session remains in the user session, the user session is terminated as well.

I assume in your case only issuing a simple logout is the way to go.

**Comment by saurabhnemade** [comment]
> > I am redirecting user to logout url but before that I am calling token endpoint to revoke both access_token and refresh_token.
> 
> @saurabhnemade - revoking the refresh token might cause the the effect you see. While this was unexpected to me as well at first, it is in the [OAuth 2.0 specs](https://datatracker.ietf.org/doc/html/rfc7009):
> 
> > If the particular
> > token is a refresh token and the authorization server supports the
> > revocation of access tokens, then the authorization server SHOULD
> > also invalidate all access tokens based on the same authorization
> > grant
> 
> When then in one browser the refresh token is revoked for that user for that client, it will remove the client session from all other user sessions of that user. If then no other client session remains in the user session, the user session is terminated as well.
> 
> I assume in your case only issuing a simple logout is the way to go.

Thanks for the clarification. I will update the implementation to just redirect to logout url. 👍 

> We are experience this problem even with an external Infinispan cluster. Could it be a variant of the #27117 that seems to be resolved in KC 24.0.2? This one was about revoking refresh tokens, but actually about an infinispan issue with the caches. We are using KC 23.0.7

Have you tried persisting sessions in infinispan itself? It allows to save sessions in postgres. We were able to get it working but we haven't deployed it on prod yet. With that we were able to successfully retain sessions after restart of keycloak and infinispan instance itself.

We hope that will solve a lot of issues as restart of keycloak server & restart of infinispan cluster will not cause loss of sessions with this strategy. As per my current analysis we majorly lose sessions when we are trying to deploy anything new which kills pods and most of users are logged out. Memory spikes also sometimes cause restarts which is also one major root cause for loss of session cache on our end.

**Comment by saurabhnemade** [comment]
> We are experience this problem even with an external Infinispan cluster. Could it be a variant of the #27117 that seems to be resolved in KC 24.0.2? This one was about revoking refresh tokens, but actually about an infinispan issue with the caches. We are using KC 23.0.7

@infl00p 
Have you tried persisting sessions in infinispan itself? It allows to save sessions in postgres. We were able to get it working but we haven't deployed it on prod yet. With that we were able to successfully retain sessions after restart of keycloak and infinispan instance itself on our testing environments.

We hope that will solve a lot of issues as restart of keycloak server & restart of infinispan cluster will not cause loss of sessions with this strategy. As per my current analysis we majorly lose sessions when we are trying to deploy anything new which kills pods and most of users are logged out. Memory spikes also sometimes cause restarts which is also one major root cause for loss of session cache on our end.

**Comment by ahus1** [comment]
> Have you tried persisting sessions in infinispan itself? It allows to save sessions in postgres.

I know the community is experimenting with this, although it was never supported by the project. See #27976 for a current attempt to make things simpler in a future release.

As it is unrelated to the main topic, please direct this discussion to one of the existing GitHub discussion. Feel free to post a link here to direct other to that discussion as well.

**Comment by Tristan971** [comment]
> the group which is formed by the Infinispan inside of Keycloak can fail, and Keycloak nodes leave unexpectedly and then re-join. When doing this, user sessions will get lost. This is expected […]

That’s fair, but if it is what happens then it doesn’t get logged. As in there were no KC logs mentionning changes in cluster view which I’d assume would be seen in such cases?

Nonetheless, (in)stability of the infinispan group (because of issues, rollouts, or scaling) is indeed the root cause in all this yes.

**Comment by Tristan971** [comment]
> There is also no simple way to merge those partitions as it is impossible to identify the entries in the caches which should remain valid, and which have been updated or expired in another partition.

I wish there was some kind of postgres-based recovery angle to this somehow. Presumably it’s not that easy though.

**Comment by ahus1** [comment]
@dglozano - here a late reply to the situation you described above:

> 4 nodes with 8 gb each and 6gb of heap size. They are utilizing ~3GB each. Cache owners is set to 4. Running Keycloak 21.1.1 too.

When preparing a setup where you have the same number of owners like nodes, you might be looking for a replicated cache instead of a distributed cache. See the work cache for an example. This might be simpler to set up and to maintain. At the same time, the documented and supported default will always be a distributed cache with 2 owners. 

> The only thing I could add is that it seems to happen whenever we get a sudden burst of calls to the refresh token endpoint. Also, is the `Failed to replace entity` errors happen both for `sessions` cache and `clientSessions` cache. MIght be worth pointing out too that we also have a Max sessions limit per user set.

Max sessions limit per user is a very expensive setting as Keycloak needs to iterate over all sessions in the cluster to find the other potential sessions of a user. I can't recommend it for high load setups.

> RIght after that burst of invalid refresh tokens and warnings from the cache, we got many of these errors and 2 of the 4 containers restarted.

The recent Keycloak HA guide describes a situation where the JGroup thread pool is exhausted and communication with the embedded Infinispan can stall, see https://www.keycloak.org/high-availability/concepts-threads. When it stalls the throughput drops, and even more requests queue up.

I recommend that you monitor the metrics on your cluster closely for that JGroups thread pool. To ensure that it doesn't exhaust, and given its default in KC24 of 200 threads, the sum of Quarkus threads over all Pods should not exceed that number. Keeping the default of 200 JGroup threads, in your setup you should limit the number of Quarkus threads to 50 as you are running 4 nodes. The lastest version of Keycloak has the parameter `http-pool-max-threads` for that. 

Once requests queue up, in earlier versions KC's liveness probes would also time out. Since KC23, the liveness-probe is non-blocking, since KC24 other probes are mostly non-blocking as well. So even when your requests queue up, the health probes will still return ok so Kubernetes will not restart your Pods due to no response to the liveness probe. You still need to ensure they have a high enough memory for a load spike ... otherwise Kubernetes will kill them due to memory.

Note that if you add custom liveness probes to your setup, it will be on you to ensure they are non-blocking.

To protect your Pod from a high load, you can limit the number of queued requests in a Pod. For that KC24 offers the parameter `http-max-queued-requests` for load shedding. Assuming a Keycloak Pod processes around 200 requests per second, a queue of 1000 would lead to maximum waiting times of around 5 seconds. When this setting is active, requests that exceed the number of queued requests will return with an HTTP 503 error. With KC24, the liveness probe with be exempt from that, as it is non-blocking it will never be queued, and therefore will not return 503 due to that. 

Given those enhancements in recent versions of Keycloak, this should resolve most of the symptoms you are experiencing.

**Comment by ahus1** [comment]
Side note to those following this issue: We're currently experimenting with user sessions stored in the database, with the option to not store them in-memory at all. See https://github.com/keycloak/keycloak/discussions/28271 for the discussion, and how to run it using the nightly build.

**Comment by ahus1** [comment]
@Tristan971 - looking at your log entries about "failing to replace" looks suspicious to me, as the warning is logged only once per request, and there are multiple requests for a single sessions probably concurrently - see the quoted snippet below. To me, this indicates multiple concurrent refresh token requests for a user session.

In the past, we've been seeing clients which misbehave in the following ways: 

* For each request, they refresh the access token - this is an anti-pattern as it puts a lot of load on Keycloak. Instead, they should contact Keycloak only when the current access token is about to expire. 

* When the access token is about to expire, they trigger a refresh request. And while this request is being processed, all parallel requests also trigger refresh requests, as they see that the access token hasn't been refreshed yet. Instead, all parallel requests should block and wait for the refresh request to complete first, and then all use the new access token from that one refresh request. 

You should be able to track down both scenarios by watching the `REFRESH_TOKEN` event. See https://www.keycloak.org/docs/latest/server_admin/#event-listener on how to show those events in the log. Those events contain the original and the next refresh token ID, so you should all information available to analyze this.

I've created a follow-up issue #28311 to collect more information for the first scenario to track the lifetimes directly so there is no need to derive them from two consecutive log entries.

````
2023-07-11 10:03:54,296 WARN  [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (executor-thread-13949) Failed to replace entity '1328551e-f813-4d60-9079-bade450008cf' in cache 'sessions'
2023-07-11 10:03:54,338 WARN  [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (executor-thread-44668) Failed to replace entity '1328551e-f813-4d60-9079-bade450008cf' in cache 'sessions'
2023-07-11 10:03:54,344 WARN  [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (executor-thread-44673) Failed to replace entity '1328551e-f813-4d60-9079-bade450008cf' in cache 'sessions'
2023-07-11 10:03:54,357 WARN  [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (executor-thread-11756) Failed to replace entity '1328551e-f813-4d60-9079-bade450008cf' in cache 'sessions'
2023-07-11 10:03:54,353 WARN  [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (executor-thread-62292) Failed to replace entity '1328551e-f813-4d60-9079-bade450008cf' in cache 'sessions'
```

**Comment by ahus1** [comment]
@rajeevBlr / @daviddelannoy - I've been looking into release 20.0.5 and found that "invalid_token" in the response might be specific to this snippet of code: 

https://github.com/keycloak/keycloak/blob/148c7695ff5b32f46f05a5f341ca5456da144987/services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java#L537-L545

This is catching an exception here, and it would not log it until you enable the TRACE logging for this class. 

In the recent version of Keycloak (since Keycloak 24), this code has moved to `RefreshTokenGrantType`, so if you're using KC24, you'll need to enable trace logging there. 

I hope the original exception will reveal more information.

**Comment by daviddelannoy** [comment]
Thanks @ahus1, I will check if these refresh token errors still occur in KC 22.x and enable logging on this class in staging environment if issue can be reproduced.

**Comment by Tristan971** [comment]
Thanks for looking into all these first @ahus1, it is plenty of helpful advice already :crossed_fingers: 

> For each request, they refresh the access token - this is an anti-pattern as it puts a lot of load on Keycloak. Instead, they should contact Keycloak only when the current access token is about to expire.

Agreed, but alas we can only control it to some extent (ratelimiting it) as we have to accomodate many 3rd-party clients of varying quality.

Either way, that explains why we see exactly 5 lines, as we do have a rate-limit exactly for that reason there, which is exactly at 5 per minute...

> And while this request is being processed, all parallel requests also trigger refresh requests [... but] should block and wait [...]

I see. This is very difficult to enforce, even from well-intentioned clients, but I'm sure we can force some request coalescing there somehow in our loadbalancers.

> I've created a follow-up issue https://github.com/keycloak/keycloak/issues/28311 to collect more information for the first scenario to track the lifetimes directly so there is no need to derive them from two consecutive log entries.

Great indeed. Looking forward to this, and hopefullly other similar extra things for the other issues you raised, so we can more easily understand when something happens that Keycloak doesn't like, then address that before suspecting bugs :sweat_smile:

**Comment by ahus1** [comment]
@Tristan971 - to add more context to the message `Failed to replace entity ... in cache 'sessions'`: This is actually Keycloak giving up after trying to write updates to cache after 25 attempts, as other concurrent writes to this session happens on every one of these writes. This means that more than 25 concurrent request (refreshes?) happen which are connect to a user session. If you have 5 attempts per client session concurrently, and 5 client sessions in a user session, this might touch this limit.  

So this makes me wonder: How many client sessions do you have connected to a user session? Do you have bursts of refreshes connected to a user session?

I've created a follow-up issue #28388 to discuss how to handle a burst of refresh requests differently.

**Comment by daviddelannoy** [comment]
Hi @ahus1,

On my side I can confirm that it is not rare at all to have 5 client sessions or more on a single user session. As many websites + mobile apps are accessible in SSO on the same realm, users can have more that 5 client sessions. Eventhough, I think it is not the most common pattern which is usually 1 to 3 max client sessions. 

They might have also multiple tabs opened in the same browser on these different clients, and some of them might be refreshed regularly and trigger refresh token. 

Actually, after some trouble (like described above) during a recent rolling upgrade with errors `Failed to replace entity ... in cache 'clientSessions'`, I can still see a few errors like this.

I can relate some  `REFRESH_TOKEN_ERROR`  to `Failed to replace entity ... in cache 'clientSessions'` 
All users impacted have several user sessions and multiple client sessions on some of these user sessions. 

In this specific case, I can see that 2 pods (clients) are making a concurrent request and got `invalid_token` error. The `refresh_token_id` provided this morning for the refresh (indeed, concurrently) is the last `updated_refresh_token_id` got yesterday : 

![image](https://github.com/keycloak/keycloak/assets/16318239/0c8749b2-8f56-4489-834e-29e5342c8af4)

**Comment by Tristan971** [comment]
> How many client sessions do you have connected to a user session? Do you have bursts of refreshes connected to a user session?

Usually only 1 or 2 that are susceptible to that (our main site and forums).
I expect this is going to be edge-case and happen only for really poorly behaved third-party clients with no retry backoff logic whatsoever.

> I've created a follow-up issue https://github.com/keycloak/keycloak/issues/28388 to discuss how to handle a burst of refresh requests differently.

Great, thanks!

**Comment by ahus1** [comment]
@daviddelannoy @Tristan971 - could please add some more information about your setups: 

* How many Keycloak nodes do you run in your cluster (Pods if it is Kubernetes)
* Which version of Keycloak are you running
* Did you configure a remote store for Keycloak (it seems it might amplify the "failed to replace entity" situation) - maybe you could share an XML snippet if you are using a custom Infinispan cache XML file.

**Comment by Tristan971** [comment]
> How many Keycloak nodes do you run in your cluster (Pods if it is Kubernetes)

Running 6 pods. Their average load is ~1 CPU core each, and ~4.3GB of effective memory use.

They run as pre-optimized container with `JAVA_OPTS=-XX:+UseShenandoahGC -Xms3g -Xmx3g -Djgroups.dns.query=keycloak-headless.keycloak.svc`

They target a 3-nodes (1 primary 2 replicas) Postgresql 15 cluster (dedicated to Keycloak), configured with 4GB of RAM per node and the following parameters:

```yaml
    parameters:
      checkpoint_completion_target: "0.9"
      default_statistics_target: "100"
      effective_cache_size: 2688MB
      effective_io_concurrency: "300"
      maintenance_work_mem: 224MB
      max_connections: "300"
      max_parallel_maintenance_workers: "2"
      max_parallel_workers: "4"
      max_parallel_workers_per_gather: "2"
      max_wal_size: 4GB
      max_worker_processes: "4"
      min_wal_size: 1GB
      random_page_cost: "1.1"
      shared_buffers: 896MB
      wal_buffers: 16MB
      work_mem: 1529kB
```

Not using PgBouncer because at the time it didn't support prepared statements, but I hear they recently added that so it could be a neat improvement to add it in I guess, since rolling restarts tend to exhaust DB connections unless performed very slowly.

> Which version of Keycloak are you running

21.1.1. It's getting a bit old, and we used to track KC upstream closely, but the requirement of mass disconnection on update made it very difficult to justify :/

> Did you configure a remote store for Keycloak

No. I actually wanted to try it because of this issue, but the lack of documentation at the time made me choose to just disable autoscaling instead.

**Comment by daviddelannoy** [comment]
> 5 client sessions in a single user session are not a problem unless those clients send a bulk of concurrent refreshes, so this shouldn't be the root cause. In your example, the difference is that it is about `clientSessions`, so a concurrent update to the client session is showing here, not a concurrent update to the user session. I don't have an explanations for that, yet. 
> 
> It might be worth noting that a user session can only have one client session per client. No sure if that client session has been replaced by a new client session in the same user session from the same client. Not sure when and how that could happen.

I guess that we probably have bulk of concurrent refreshes. I think your event listener will help to identify rogue clients  more easily than checking events and timestamps.

**Comment by daviddelannoy** [comment]
> @daviddelannoy @Tristan971 - could please add some more information about your setups: 
> 
> * How many Keycloak nodes do you run in your cluster (Pods if it is Kubernetes)

3 nodes behind a load balancer. distributed-cache with 2 owners for sessions and clientSessions, using JDBC_PING for jgroups with the postgresql database (totally reliable from our experience). 

> * Which version of Keycloak are you running

22.0.5, upgrade to 24 in coming months.
Exactly like @Tristan971, we have the same requirement about keeping users logged in. 
So when there are major updates/changes of infinispan dependency and session stored cannot be unmarshalled with the new version, it is also not easy to justify. 

> * Did you configure a remote store for Keycloak (it seems it might amplify the "failed to replace entity" situation) - maybe you could share an XML snippet if you are using a custom Infinispan cache XML file.
> 

Yes but as said before, not with an external ispn cluster but directly to a postgres database (see #10803). 
Here is a snippet of the persistence part. It should work now without any custom module to override `CacheDecorators` since the work you have done on #24774 embedded in KC23. 

```
<distributed-cache name="clientSessions" owners="2">
            <state-transfer enabled="true" await-initial-transfer="false" />
            <expiration />
            <persistence>
                <string-keyed-jdbc-store xmlns="urn:infinispan:config:store:jdbc:14.0"
                    dialect="POSTGRES" shared="true" passivation="false" purge="false" preload="false">
                    <connection-pool
                        properties-file="/path/to/keycloak/conf/ispn-agroal.properties" />
                    <string-keyed-table create-on-start="true" prefix="ispn">
                        <id-column name="id" type="VARCHAR(255)" />
                        <data-column name="datum" type="BYTEA" />
                        <timestamp-column name="timestamp" type="BIGINT" />
                        <segment-column name="segment" type="INT" />
                    </string-keyed-table>
                </string-keyed-jdbc-store>
            </persistence>
        </distributed-cache>
```

`await-initial-transfer="false"` works fine to start the instance quickly and load the cache asynchronously, however as you already mentioned in the issue/PR about detecting remote store, it is far from perfect in term of admin experience, for example in the admin console where all sessions are not visible immediately. 
I think that it is better to limit the number of cache entries, reduce the heap size, preload everything and store exceeding sessions into the database. 

ISPN errors about writing/replacing entries are not related to this option, it was usually worse before with the full preload

**Comment by ahus1** [comment]
Added a follow-up issue to add details to the log messages #28429

**Comment by ahus1** [comment]
For the refresh token errors, there will be additional information logged why this is the case with #28429. Failed refreshes can have multiple causes like for example a locked user account, and we'll see it in the logs. With all events containing the user session ID, it will be simpler to correlate them. This will be available in KC25, and I'll also backport to KC24 and it will be eventually available in a patch release. 

We don't do community releases, so if you want to use it in an earlier release of Keycloak, you will need to backport this yourself and build your own image.

**Comment by ahus1** [comment]
Created a spin-off issue for the "Failed to replace entity" message here #28466

**Comment by ahus1** [comment]
@daviddelannoy / @Tristan971 - one more question about your environments: 

* How do you manage your pods: as a deployment or a stateful set at the moment?

**Comment by Tristan971** [comment]
Deployment, in my case. And I really need to replace that with STS, though the issue triggered in situations I think weren't really related. ie:
- without rollouts (ie the CPU/network strain events would make pods go unhealthy but not kill  them)
- with 1-by-1 replacements (autoscaling events are slow as our traffic isn't spiky at all)

**Comment by daviddelannoy** [comment]
> @daviddelannoy / @Tristan971 - one more question about your environments:
> 
> * How do you manage your pods: as a deployment or a stateful set at the moment?

Our environment is not in a k8s cluster, only 3 simples EC2 instances in a target group behind a load balancer, using auto scaling group for rolling updates. 
That is quite an interesting and important point in fact : with @Tristan971  we have 2 totally different deployment environments (his is k8s and not persisting sessions, ours is not k8s and persisting sessions in DB), nevertheless we experimented exactly the same kind of replace errors in ISPN cache when doing rolling updates. 

I don't remember if I mentioned it before, but from what I have observed, replace errors will occur more frequently and KC response times increase **when a server is removed from the cluster**, not when adding a new one. It seems logical as our new instances load the cache asynchronously (so replication of entries takes time), but when removing  and instance there much more sessions to rebalance to other nodes.

**Comment by ahus1** [comment]
I'm closing this issue as all behaviors described here have been explained in the comments, worked or worked in spin-off tickets, especially #28429 (which is about to be backported to KC24) and #28466 (which will require additional debug logs). Therefore closing this issue.

**Comment by CosminPerRam** [comment]
Hey, I've stumbled across this issue when I was investigating the same problem, I was receiving a refresh token and upon trying to update it I would get "Token is not active", turns out that on our configuration we didn't included the `offline_access` scope by default and after the SSO max timeout we would encounter this issue.

**Comment by souravs17031999** [comment]
@CosminPerRam Does this `offline_access` is really your requirement ?, remember offline_access is usually not required for normal user authorization usecases and only used when client application needs to perform some actions on behalf of user even after user logged out. Offline access if not used properly can be misused especially considering security of offline tokens if leaked can lead to attackers misuing it for minting newer tokens.
